graphql_compiler/interpreter/immutable_stack.py

```diff
@@ -0,0 +1,35 @@
+# Copyright 2020-present Kensho Technologies, LLC.
+from dataclasses import dataclass
+from typing import Any, Optional, Tuple
+
+
+@dataclass(frozen=True)
+class ImmutableStack:
+    """An immutable stack of arbitrary (heterogeneously-typed) values.
+
+    Specifically designed for cheap structural sharing, in order to avoid deep copies or
+    bugs caused by mutations of shared data.
+    """
+
+    # The following attributes are considered visible and safe for direct external use.
+    value: Any  # The value contained within this stack node.
+    depth: int  # The number of stack nodes contained in the tail of the stack.
+    tail: Optional["ImmutableStack"]  # The node that represents the rest of the stack, if any.
```

Comment:
You can leave out `depth` from the constructor to avoid invalid stacks like `ImutableStack(5, 8, None)`.

Reply:
Yeah good point 👍

Reply:
My initial thought was that nobody should be instantiating `ImmutableStack` directly using the constructor, but hidden constructors aren't really a thing in Python...